drivers/sensor/adi/adxl367/adxl367_stream.c

```diff
@@ -0,0 +1,558 @@
+		}
+	} while (cqe != NULL);
+
+	/* Bail/cancel attempt to read sensor on any error */
+	if (res != 0) {
+		adxl367_sqe_done(cfg, current_sqe, res);
+		return;
+	}
+
+	enum sensor_stream_data_opt data_opt;
+
+	if ((fifo_wmark_cfg != NULL) && (fifo_full_cfg == NULL)) {
+		data_opt = fifo_wmark_cfg->opt;
+	} else if ((fifo_wmark_cfg == NULL) && (fifo_full_cfg != NULL)) {
+		data_opt = fifo_full_cfg->opt;
+	} else {
+		data_opt = MIN(fifo_wmark_cfg->opt, fifo_full_cfg->opt);
+	}
+
+	if (data_opt == SENSOR_STREAM_DATA_NOP || data_opt == SENSOR_STREAM_DATA_DROP) {
+		uint8_t *buf;
+		uint32_t buf_len;
+
+		/* Clear streaming_sqe since we're done with the call */
+		data->sqe = NULL;
+		if (rtio_sqe_rx_buf(current_sqe, sizeof(struct adxl367_fifo_data),
+				    sizeof(struct adxl367_fifo_data), &buf, &buf_len) != 0) {
+			adxl367_sqe_done(cfg, current_sqe, -ENOMEM);
+			return;
+		}
+
+		struct adxl367_fifo_data *rx_data = (struct adxl367_fifo_data *)buf;
+
+		memset(buf, 0, buf_len);
+		rx_data->is_fifo = 1;
+		rx_data->timestamp = data->timestamp;
+		rx_data->int_status = status;
+		rx_data->fifo_byte_count = 0;
+
+		if (data_opt == SENSOR_STREAM_DATA_DROP) {
+			/* Flush the FIFO by disabling it. Save current mode for after the reset. */
+			adxl367_fifo_flush_rtio(dev);
+			return;
+		}
+
+		adxl367_sqe_done(cfg, current_sqe, 0);
+		return;
+	}
+
+	struct rtio_sqe *write_fifo_addr = rtio_sqe_acquire(data->rtio_ctx);
+	struct rtio_sqe *read_fifo_data = rtio_sqe_acquire(data->rtio_ctx);
+	struct rtio_sqe *complete_op = rtio_sqe_acquire(data->rtio_ctx);
+	const uint8_t reg[2] = {ADXL367_SPI_READ_REG, ADXL367_FIFO_ENTRIES_L};
+
+	rtio_sqe_prep_tiny_write(write_fifo_addr, data->iodev, RTIO_PRIO_NORM, reg, 2, NULL);
+	write_fifo_addr->flags = RTIO_SQE_TRANSACTION;
+	rtio_sqe_prep_read(read_fifo_data, data->iodev, RTIO_PRIO_NORM, data->fifo_ent, 2,
+						current_sqe);
+	read_fifo_data->flags = RTIO_SQE_CHAINED;
+	rtio_sqe_prep_callback(complete_op, adxl367_process_fifo_samples_cb, (void *)dev,
+							current_sqe);
+
+	rtio_submit(data->rtio_ctx, 0);
+}
+
+void adxl367_stream_irq_handler(const struct device *dev)
+{
+	struct adxl367_data *data = (struct adxl367_data *) dev->data;
+
+	if (data->sqe == NULL) {
+		return;
+	}
+
+	data->timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+
+	struct rtio_sqe *write_status_addr = rtio_sqe_acquire(data->rtio_ctx);
+	struct rtio_sqe *read_status_reg = rtio_sqe_acquire(data->rtio_ctx);
+	struct rtio_sqe *check_status_reg = rtio_sqe_acquire(data->rtio_ctx);
+	const uint8_t reg[2] = {ADXL367_SPI_READ_REG, ADXL367_STATUS};
+
+	rtio_sqe_prep_tiny_write(write_status_addr, data->iodev, RTIO_PRIO_NORM, reg, 2, NULL);
+	write_status_addr->flags = RTIO_SQE_TRANSACTION;
+	rtio_sqe_prep_read(read_status_reg, data->iodev, RTIO_PRIO_NORM, &data->status, 1, NULL);
+	read_status_reg->flags = RTIO_SQE_CHAINED;
+	rtio_sqe_prep_callback(check_status_reg, adxl367_process_status_cb, (void *)dev, NULL);
```

Comment:
Would recommend using rtio_sqe_prep_callback_no_cqe for all of these to avoid creating a completion that you don't actually care about, but not a blocker if the driver is working